File: app/src/main/java/com/soundcloud/android/playback/playqueue/PlayQueueOperations.java
package com.soundcloud.android.playback.playqueue;

import static com.soundcloud.android.ApplicationModule.RX_HIGH_PRIORITY;
import static com.soundcloud.java.collections.Lists.transform;
import static com.soundcloud.java.collections.ListsFunctions.cast;

import com.soundcloud.android.model.Urn;
import com.soundcloud.android.playback.PlayQueueItem;
import com.soundcloud.android.playback.PlayQueueManager;
import com.soundcloud.android.playback.PlayQueueStorage;
import com.soundcloud.android.playback.TrackQueueItem;
import com.soundcloud.android.tracks.TrackItem;
import com.soundcloud.android.tracks.TrackItemRepository;
import com.soundcloud.android.utils.DiffUtils;
import com.soundcloud.java.collections.Lists;
import io.reactivex.Scheduler;
import io.reactivex.Single;
import io.reactivex.functions.Function;

import javax.inject.Inject;
import javax.inject.Named;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class PlayQueueOperations {

    private final Scheduler scheduler;
    private final PlayQueueManager playQueueManager;
    private final TrackItemRepository trackItemRepository;
    private final PlayQueueStorage playQueueStorage;

    @Inject
    public PlayQueueOperations(@Named(RX_HIGH_PRIORITY) Scheduler scheduler,
                               PlayQueueManager playQueueManager,
                               TrackItemRepository trackItemRepository,
                               PlayQueueStorage playQueueStorage) {
        this.scheduler = scheduler;
        this.playQueueManager = playQueueManager;
        this.trackItemRepository = trackItemRepository;
        this.playQueueStorage = playQueueStorage;
    }

    public Single<List<TrackAndPlayQueueItem>> getTracks() {
        return Single.defer(this::loadTracks);
    }

    Single<Map<Urn, String>> getContextTitles() {
        return playQueueStorage.contextTitles().subscribeOn(scheduler);
    }

    private Single<List<TrackAndPlayQueueItem>> loadTracks() {
        final List<TrackQueueItem> playQueueItems = Lists.transform(playQueueManager.getPlayQueueItems(input -> input != null && input.isTrack() && input.getUrn().isTrack()),
                                                                    cast(TrackQueueItem.class));

        final Function<Map<Urn, TrackItem>, List<TrackAndPlayQueueItem>> fulfillWithKnownProperties = urnTrackMap -> toTrackAndPlayQueueItem(playQueueItems, urnTrackMap);
        final List<Urn> uniqueTrackUrns = DiffUtils.deduplicate(transform(playQueueItems, PlayQueueItem.TO_URN));

        return trackItemRepository
                .fromUrns(uniqueTrackUrns)
                .map(fulfillWithKnownProperties)
                .subscribeOn(scheduler);
    }

    private List<TrackAndPlayQueueItem> toTrackAndPlayQueueItem(List<TrackQueueItem> playQueueItems,
                                                                Map<Urn, TrackItem> knownProperties) {
        final ArrayList<TrackAndPlayQueueItem> trackItems = new ArrayList<>(playQueueItems.size());

        for (TrackQueueItem item : playQueueItems) {
            addTrackAndPlayQueueItemIfPresent(knownProperties, trackItems, item);
        }
        return trackItems;
    }

    private void addTrackAndPlayQueueItemIfPresent(Map<Urn, TrackItem> urnTrackMap,
                                                   ArrayList<TrackAndPlayQueueItem> trackItems,
                                                   TrackQueueItem item) {
        final Urn urn = item.getUrn();
        if (urnTrackMap.containsKey(urn)) {
            trackItems.add(new TrackAndPlayQueueItem(urnTrackMap.get(urn), item));
        }
    }
}

File: app/src/main/java/com/soundcloud/android/playback/playqueue/PlayQueueOperations.kt
package com.soundcloud.android.playback.playqueue

import com.soundcloud.android.ApplicationModule.RX_HIGH_PRIORITY
import com.soundcloud.android.model.Urn
import com.soundcloud.android.playback.PlayQueueManager
import com.soundcloud.android.playback.PlayQueueStorage
import com.soundcloud.android.playback.TrackQueueItem
import com.soundcloud.android.playlists.PlaylistRepository
import com.soundcloud.android.stations.StationsRepository
import com.soundcloud.android.tracks.TrackItem
import com.soundcloud.android.tracks.TrackItemRepository
import com.soundcloud.android.tracks.TrackRepository
import com.soundcloud.android.users.UserRepository
import io.reactivex.Scheduler
import io.reactivex.Single
import io.reactivex.functions.Function4
import javax.inject.Inject
import javax.inject.Named

open class PlayQueueOperations
@Inject
constructor(@param:Named(RX_HIGH_PRIORITY) private val scheduler: Scheduler,
            private val playQueueManager: PlayQueueManager,
            private val trackItemRepository: TrackItemRepository,
            private val playQueueStorage: PlayQueueStorage,
            private val userRepository: UserRepository,
            private val stationsRepository: StationsRepository,
            private val playlistRepository: PlaylistRepository,
            private val trackRepository: TrackRepository) {

    open val tracks: Single<List<TrackAndPlayQueueItem>>
        get() = Single.defer { this.loadTracks() }

    open val contextTitles: Single<Map<Urn, String>>
        get() {
            val contextUrns = playQueueStorage.contextUrns

            val stations = getStations(contextUrns) { it.isStation }
            val users = getUsers(contextUrns) { it.isUser }
            val playlists = getPlaylists(contextUrns) { it.isPlaylist }
            val tracks = getTracks(contextUrns) { it.isTrack }

            return Single.zip<Map<Urn, String>, Map<Urn, String>, Map<Urn, String>, Map<Urn, String>, Map<Urn, String>>(users,
                    stations,
                    playlists,
                    tracks,
                    Function4 { t1, t2, t3, t4 -> t1.plus(t2).plus(t3).plus(t4) }).subscribeOn(scheduler)
        }

    private fun getStations(contextUrns: MutableList<Urn>, function: (Urn) -> Boolean): Single<Map<Urn, String>> {
        if (contextUrns.filter(function).isNotEmpty()) {
            return stationsRepository
                    .stationsMetadata(contextUrns.filter(function))
                    .map<Map<Urn, String>> { it.associateBy({ it.urn() }, { it.title() }) }

        } else {
            return Single.just(emptyMap())
        }
    }

    private fun getUsers(contextUrns: MutableList<Urn>, function: (Urn) -> Boolean): Single<Map<Urn, String>> {
        if (contextUrns.filter(function).isNotEmpty()) {
            return userRepository
                    .usersInfo(contextUrns.filter(function))
                    .map<Map<Urn, String>> { it.associateBy({ it.urn() }, { it.username() }) }

        } else {
            return Single.just(emptyMap())
        }
    }

    private fun getPlaylists(contextUrns: MutableList<Urn>, function: (Urn) -> Boolean): Single<Map<Urn, String>>? {
        if (contextUrns.filter(function).isNotEmpty()) {
            return playlistRepository
                    .withUrns(contextUrns.filter(function))
                    .map<Map<Urn, String>> { it.entries.associate { it.key to it.value.title() } }

        } else {
            return Single.just(emptyMap())
        }
    }

    private fun getTracks(contextUrns: MutableList<Urn>, function: (Urn) -> Boolean): Single<Map<Urn, String>>? {
        if (contextUrns.filter(function).isNotEmpty()) {
            return trackRepository
                    .fromUrns(contextUrns.filter(function))
                    .map<Map<Urn, String>> { it.entries.associate { it.key to it.value.title() } }

        } else {
            return Single.just(emptyMap())
        }
    }

    private fun loadTracks(): Single<List<TrackAndPlayQueueItem>> {
        val playQueueItems = playQueueManager
                .getPlayQueueItems { it != null && it.urn.isTrack }
                .map { it as TrackQueueItem }

        val uniqueTrackUrns = playQueueItems
                .map { it.urn }
                .distinct()

        return trackItemRepository
                .fromUrns(uniqueTrackUrns)
                .map { toTrackAndPlayQueueItem(playQueueItems, it) }
                .subscribeOn(scheduler)
    }

    private fun toTrackAndPlayQueueItem(playQueueItems: List<TrackQueueItem>,
                                        knownProperties: Map<Urn, TrackItem>): List<TrackAndPlayQueueItem> {
        return playQueueItems
                .filter { knownProperties.containsKey(it.urn) }
                .map { TrackAndPlayQueueItem(knownProperties[it.urn], it) }
    }

}
